src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -954,25 +963,24 @@ void ObjectMonitor::ReenterI(JavaThread* current, ObjectWaiter* currentNode) {
     if (TryLock(current) > 0) break;
     if (TrySpin(current) > 0) break;
 
-    // State transition wrappers around park() ...
-    // ReenterI() wisely defers state transitions until
-    // it's clear we must park the thread.
     {
       OSThreadContendState osts(current->osthread());
-      ThreadBlockInVM tbivm(current);
 
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
-      current->set_suspend_equivalent();
-      current->_ParkEvent->park();
+      assert(current->thread_state() == _thread_in_vm, "invariant");
 
-      // were we externally suspended while we were waiting?
-      for (;;) {
-        if (!current->handle_special_suspend_equivalent_condition()) break;
-        if (_succ == current) { _succ = NULL; OrderAccess::fence(); }
-        current->java_suspend_self();
-        current->set_suspend_equivalent();
+      current->frame_anchor()->make_walkable(current);
+      OrderAccess::storestore();
```

Comment:
Repeating this comment since my original is marked resolved, but I'm not seeing
a new comment here. I originally pointed this out in the resolved conversation,
but that doesn't make the comment show up in the review emails.

Needs a comment explaining what the memory sync is for.

Reply:
I seem to have fixed 2 of out 3.